### PR TITLE
fancybox addon v.1.03

### LINKED ISSUE #12331: [fancybox addon] Image gallery is created for only the images <= clicked image
- [x ] I have searched open and closed issues for duplicates

### Bug Description

When you click on an image in a post with images and this image is not the last one attached, the fancybox gallery lets you only iterate through the images prior and including the one you clicked on. Images which are attached after the one you clicked are not included in the gallery. 

### Steps to Reproduce

1.  Visit a post with multiple images attached (and have fancybox addon activated), e.g. [here](https://friendica.xyz/display/adf174d5-1363-8cb0-4dcd-455672791341)
2.  Click on the first, second or third image

Actual Result:

The upcoming fancybox gallery lets you only cycle through the one (if you clicked on the first image), two (if you clicked the second image), or three images (if you clicked the third image). 

Expected Result:

Regardless which images on a multiple image attachment post you clicked, the gallery should include all images from the attachment for cycling through. 

### Platform Info

Friendica Version: 'Giant Rhubarb' 2022.12-dev - 1496

Friendica Source: https://github.com/friendica/friendica/tree/62f982cf7e2cc976c578e36c0244d79aeb705522

PHP version: 7.4.33

SQL version: 10.6.11-MariaDB-0ubuntu0.22.04.1

### PR DESCRIPTION
**Pull request by brockhaus**
This PR removes the lightbox hook out of the addon and replaces all URLs on images inlined in posts with the fancybox hook.
URLs in divs class="type-link" won't be touched, as these URLs have pages not images linked to.

Fix https://github.com/friendica/friendica/issues/12331

### CODE REVIEW

File: fancybox/asset/fancybox/fancybox.config.js [hunk 1]
```diff
@@ -1,3 +1,4 @@
 $(document).ready(function() {
     $.fancybox.defaults.loop = "true";
+    $("body").off("click", ".wall-item-body a img");
```
Comment (MrPetovan) [inline fancybox/asset/fancybox/fancybox.config.js:4]: Please add a reference to the listener it deactivates, ideally the file and the line.

File: fancybox/fancybox.php [hunk 2]
```diff
@@ -28,27 +28,43 @@ function fancybox_footer(App $a, string &$str)
...
-	$html = $b['html'];
-	while (preg_match($pattern, $html, $matches, PREG_OFFSET_CAPTURE)) {
-		if (is_array($matches)) {
-			$matches = $matches[0];
-		}
-		$part     = $matches[0];
-		$replaced = str_replace('<a href', '<a data-fancybox="' . $gallery . '" href', $part);
-		$replaced = str_replace('<div class="body-attach"', '<div class="body-attach done"', $replaced);
-		$html     = str_replace($part, $replaced, $html);
+function fancybox_render($a, array &$b){
```
Comment (MrPetovan) [inline fancybox/fancybox.php]: Why did you remove the type hint of the first parameter?
```suggestion
function fancybox_render(App $a, array &$b){
```

File: fancybox/fancybox.php [hunk 3]
```diff
@@ -28,27 +28,43 @@ function fancybox_footer(App $a, string &$str)
...
+	$gallery = 'gallery-' . $b['item']['uri-id'] ?? random_int(1000000, 10000000);
+
+	// prevent urls in <div class="type-link"> to be replaced
+	$b['html'] = preg_replace_callback(
+		'#<div class="type-link">.*?</div>#s',
+		function ($matches) use ($gallery) {
+			return str_replace('<a href', '<a data-nofancybox="" href', $matches[0]);
+		},
+		$b['html']
+	);
```
Comment (MrPetovan) [inline fancybox/fancybox.php]: This can be done with an existing Friendica method called `Util\Strings::performWithEscapedBlocks` with the same regular expression you came up with.

It would look like this (untested):
```php
	$b['html'] = \Friendica\Util\Strings::performWithEscapedBlocks(
		$b['html'],
		'#<div class="type-link">.*?</div>#s',
		function ($text) use ($gallery) {
			// This processes images inlined in posts
			// Frio / Vier hooks für lightbox are un-hooked in fancybox-config.js. So this works for them, too!
			//if (!in_array($a->getCurrentTheme(),['vier','frio']))
			// normal post inline linked images
			$text = preg_replace(
				'#<a[^>]*href="([^"]*)"[^>]*>(<img[^>]*src="[^"]*"[^>]*>)</a>#',
				'<a data-fancybox="' . $gallery . '" href="$1">$2</a>',
				$text
			);

			// Local content images attached:
			return preg_replace_callback(
				'#<div class="body-attach">.*?</div>#s',
				function ($matches) use ($gallery) {
					return str_replace('<a href', '<a data-fancybox="' . $gallery . '" href', $matches[0]);
				},
				$text
			);
		}
	);
```
Notice the change from `preg_replace_callback` to `preg_replace` since we don't need to care about `no-fancybox` elements.

**Review by MrPetovan — CHANGES_REQUESTED** [review]

File: fancybox/asset/fancybox/fancybox.config.js (hunk 1 above)
Comment (brockhaus) [reply fancybox/asset/fancybox/fancybox.config.js:4]: Added comment for next PR

File: fancybox/fancybox.php (hunk 2 above)
Comment (brockhaus) [reply fancybox/fancybox.php]: This was because of test code. I forget to restore it.

File: fancybox/fancybox.php (hunk 3 above)
Comment (brockhaus) [reply fancybox/fancybox.php]: Please elaborate what this method does (or send me a link to the description). W/o that I don't know how this suggested code change should work?

Comment (brockhaus) [reply fancybox/fancybox.php]: Ah.. I see in the comment description of that function in code.. Hmm.. Have to check that..

Comment (MrPetovan) [reply fancybox/fancybox.php]: Were you able to test?

Comment (brockhaus) [reply fancybox/fancybox.php]: Yes I tested this. It worked like intended. So I think the (not yet known to me) method does what it is supposed to do.
Img URLs in "type-link" divs are not touched. :-)

Comment (MrPetovan) [reply fancybox/fancybox.php]: I ran into the same exact use case you did and decided to make my future life and the future life of everyone who contributes to the project easier!

**Review by MrPetovan — APPROVED** [review]
